contracts/RouterStrategy.sol

```diff
@@ -0,0 +1,232 @@
+        internal
+        override
+        returns (uint256 _liquidatedAmount, uint256 _loss)
+    {
+        uint256 balance = balanceOfWant();
+        if (balance >= _amountNeeded) {
+            return (_amountNeeded, 0);
+        }
+
+        uint256 toWithdraw = _amountNeeded.sub(balance);
+        _withdrawFromYVault(toWithdraw);
+
+        uint256 totalAssets = balanceOfWant();
+        if (_amountNeeded > totalAssets) {
+            _liquidatedAmount = totalAssets;
+            _loss = _amountNeeded.sub(totalAssets);
+        } else {
+            _liquidatedAmount = _amountNeeded;
+        }
+    }
+
+    function _withdrawFromYVault(uint256 _amount) internal {
+        if (_amount == 0) {
+            return;
+        }
+
+        uint256 _balanceOfYShares = yVault.balanceOf(address(this));
+        uint256 sharesToWithdraw =
+            Math.min(_investmentTokenToYShares(_amount), _balanceOfYShares);
+
+        if (sharesToWithdraw == 0) {
+            return;
+        }
+
+        yVault.withdraw(sharesToWithdraw, address(this), maxLoss);
+    }
+
+    function liquidateAllPositions()
+        internal
+        override
+        returns (uint256 _amountFreed)
+    {
+        return
+            yVault.withdraw(
+                yVault.balanceOf(address(this)),
+                address(this),
+                maxLoss
+            );
+    }
+
+    function prepareMigration(address _newStrategy) internal override {
+        IERC20(yVault).safeTransfer(
+            _newStrategy,
+            IERC20(yVault).balanceOf(address(this))
+        );
+    }
+
+    function protectedTokens()
+        internal
+        view
+        override
+        returns (address[] memory ret)
+    {
+        ret = new address[](1);
+        ret[0] = address(yVault);
+        return ret;
+    }
+
+    function ethToWant(uint256 _amtInWei)
+        public
+        view
+        virtual
+        override
+        returns (uint256)
+    {
+        return _amtInWei;
+    }
+
+    function setMaxLoss(uint256 _maxLoss) public onlyEmergencyAuthorized {
+        maxLoss = _maxLoss;
+    }
+
+    function _checkAllowance(
+        address _contract,
+        address _token,
+        uint256 _amount
+    ) internal {
+        if (IERC20(_token).allowance(address(this), _contract) < _amount) {
+            IERC20(_token).safeApprove(_contract, 0);
+            IERC20(_token).safeApprove(_contract, type(uint256).max);
+        }
+    }
+
+    function balanceOfWant() public view returns (uint256) {
+        return want.balanceOf(address(this));
+    }
+
+    function _investmentTokenToYShares(uint256 amount)
+        internal
+        view
+        returns (uint256)
+    {
+        return amount.mul(10**yVault.decimals()).div(yVault.pricePerShare());
+    }
+
+    function valueOfInvestment() public view returns (uint256) {
+        return
+            yVault.balanceOf(address(this)).mul(yVault.pricePerShare()).div(
```

Comment:
I guess its just a comment, Depending on the version of the vault pricePerShare it may be beneficial to wait for harvest to go past the locked profit release window after a harvest

Reply:
that's an interesting protection we should do. make sure sleeping 8hs before the tx doesn't affect the pps.